utils/tail/tail.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tail implements "tail -F" functionality following rotated logs
+package tail
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/exp/inotify"
+)
+
+type Tail struct {
+	reader   *bufio.Reader
+	filename string
+	file     *os.File
+	stop     chan bool
+	watcher  *inotify.Watcher
+}
+
+const retryReopenInterval = time.Second
+const maxReopenAttempts = 3
+
+// NewTail starts opens the given file and watches it for deletion/rotation
+func NewTail(filename string) (t *Tail, err error) {
+	t = &Tail{
+		filename: filename,
+	}
+	t.file, err = os.Open(t.filename)
+	if err != nil {
+		glog.Errorf("Open failed on %s: %v", t.filename, err)
+		return nil, err
+	}
+	// TODO: not interested in old events?
+	//t.file.Seek(0, os.SEEK_END)
+	t.reader = bufio.NewReader(t.file)
+	t.stop = make(chan bool)
+	t.watcher, err = inotify.NewWatcher()
+	if err != nil {
+		glog.Errorf("Inotify init failed on %s: %v", t.filename, err)
+		return
+	}
+	go t.watchFile()
+	return t, nil
+}
+
+// Read implements the io.Reader interface for Tail
+func (t *Tail) Read(p []byte) (n int, err error) {
+	n, err = t.reader.Read(p)
+	if err != nil {
+		return 0, io.EOF
```

Comment:
EOF should be used to signal there is no more data, but in this case it is used to signal that `readLinesFromFile` should retry. I think it would be better to just return 0 bytes read and nil error in this case (unless `err` is an actual error and not EOF). The logic in `readLinesFromFile`  will need to be updated too.